### ISSUE TITLE
not enough points for tracking

### ISSUE THREAD

**Issue by Warsin** [CIFASIS/dense-sptam#3]
when i launch the kitti.launch and the kitti_04.bag which i download from the sptam project, i meet this　warning , and then i have no output result in ~/.ros/clouds

**Comment by Warsin** [comment]
when i run 
 roslaunch src/dense-sptam/launch/kitti.launch &
rosbag play --clock path/to/your/kitti.bag 
i meet this in terminal:
[ INFO] [1529570960.970479241]: Loading nodelet /dense of type dense/denseNodelet to manager nodelet_manager with the following remappings:
[ INFO] [1529570960.971035005]: /dense/keyframes -> /sptam/keyframes
[ INFO] [1529570961.269934397]: S-PTAM stereo node initialized.
[ INFO] [1529570961.630599970]: DENSE node initialized.
rosbag play --clock kitti_04.bag 
[ INFO] [1529570976.885650110]: Opening kitti_04.bag

Waiting 0.2 seconds after advertising topics... done.

Hit space to toggle paused, or 's' to step.
[ INFO] [1529570977.674896351, 1385799339.561161995]: init calib8.108940               
[ INFO] [1529570977.675609345, 1385799339.561161995]: baseline: 0.537151
[ INFO] [1529570977.720253008, 1385799339.561161995]: Trying to intialize map...
[ INFO] [1529570977.787974284, 1385799339.561161995]: Map initialized with 130 points.
[ WARN] [1529570980.442324406, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570981.884402515, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570983.413479241, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570983.759056404, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570985.055157351, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570985.549257729, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570985.766052891, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570986.231529569, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570986.975968928, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570987.655287206, 1385799339.561161995]: Not enough points for tracking.
[ WARN] [1529570988.028875621, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570989.300185360, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570989.997125320, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570991.312939734, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570991.565123562, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570992.722895759, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570992.765493756, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570993.143247701, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570993.951083267, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570995.041895397, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570995.827613300, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570995.968390639, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570996.166484334, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570996.441586187, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570996.687139098, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570996.949713367, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570997.856295341, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570998.377672408, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570999.179960507, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570999.243885647, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529570999.540931539, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571000.121260339, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571001.125941498, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571001.476573238, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571002.378970756, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571002.725887595, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571003.082918782, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571003.305880644, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571003.679541904, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571004.271051863, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571004.490242222, 1385799339.561161995]: Not enough points for tracking.
ERROR: There is not any covisibility keyframes observing the tracked points
[ WARN] [1529571004.765001701, 1385799339.561161995]: Not enough points for tracking

**Comment by Warsin** [comment]
but when i run
roslaunch src/dense-sptam/launch/kitti.launch &
rosbag play --clock --rate 0.1 kitti_04.bag
i meet this:
auto-starting new master
process[master]: started with pid [85]
ROS_MASTER_URI=http://127.0.0.1:11311

setting /run_id to b1117b30-7532-11e8-a2c8-000c2986eb06
process[rosout-1]: started with pid [98]
started core service [/rosout]
process[start_broadcaster-2]: started with pid [118]
process[sptam_path-3]: started with pid [119]
process[nodelet_manager-4]: started with pid [137]
[ INFO] [1529572125.428744450]: Initializing nodelet with 4 worker threads.
process[sptam-5]: started with pid [161]
[ INFO] [1529572125.893189418]: Loading nodelet /sptam of type sptam/sptam_nodelet to manager nodelet_manager with the following remappings:
[ INFO] [1529572125.893319381]: /stereo/left/camera_info -> /kitti_stereo/left/camera_info
[ INFO] [1529572125.893357602]: /stereo/left/image_rect -> /kitti_stereo/left/image_rect
[ INFO] [1529572125.893587453]: /stereo/right/camera_info -> /kitti_stereo/right/camera_info
[ INFO] [1529572125.893632843]: /stereo/right/image_rect -> /kitti_stereo/right/image_rect
  crossCheck: 0
OpenCV: threads set to -1
  nfeatures: 1000
  qualityLevel: 0.01
  minDistance: 15
  blockSize: 3
  useHarrisDetector: 0
  k: 0.04
  nfeatures: 1000
  qualityLevel: 0.01
  minDistance: 15
  blockSize: 3
  useHarrisDetector: 0
  k: 0.04
  bytes: 32
  use_orientation: 0
  bytes: 32
  use_orientation: 0
process[dense-6]: started with pid [188]
[ INFO] [1529572126.490738690]: Loading nodelet /dense of type dense/denseNodelet to manager nodelet_manager with the following remappings:
[ INFO] [1529572126.490883237]: /dense/keyframes -> /sptam/keyframes
[ INFO] [1529572126.782521508]: S-PTAM stereo node initialized.
[ INFO] [1529572127.322514652]: DENSE node initialized.
[ INFO] [1529572130.203967090]: Opening kitti_04.bag

Waiting 0.2 seconds after advertising topics... done.

Hit space to toggle paused, or 's' to step.
[ INFO] [1529572130.735223612, 1385799339.661297826]: init calib8.108940               
[ INFO] [1529572130.735660508, 1385799339.661297826]: baseline: 0.537151
[ INFO] [1529572130.785293238, 1385799339.661297826]: Trying to intialize map...
[ INFO] [1529572130.842278877, 1385799339.661297826]: Map initialized with 137 points.
 [RUNNING]  Bag Time: 1385799367.148107   Duration: 27.586945 / 28.108940               
Done.
it seems that just play the bag and do nothing.

**Comment by taihup** [comment]
Are you using rviz to visualise the output?
Try to check if the topics are being published using the terminal command
rostopic list and rostopic echo <topic_name>

**Comment by Warsin** [comment]
@taihup oh~,there are some topics are being published,"/sptam/keyframes  ,/sptam/local_keyframes,   /sptam/local_map, /sptam/robot/path, /sptam/robot/pose " and so on ,but the topic /dense/dense_cloud is not being published.
and the result is also not saved

root@ubuntu-virtual-machine:/usr/src/dense_sptam# rostopic echo /dense/dense_cloud
WARNING: no messages received and simulated time is active.
Is /clock being published?

**Comment by taihup** [comment]
you must create the directory clouds inside the ~/.ros/ directory. This is:
cd ~/.ros
mkdir clouds

**Comment by Warsin** [comment]
of course ,i have created the directory clouds,but it did not save anything.  Is this parameter"output_dir" should do some change?

**Comment by Warsin** [comment]
@taihup hi,i am so sorry  for interrupting you，i found it not normally do these code, it did not do the save_all() rightly . can you tell me the right way to stop the project , i used crtl+c to stop it ,was it right?
dense::denseInterface::~denseInterface()
{
	std::cout << "Starting DENSE node cleanup..." << std::endl;
	this->dense_->point_clouds_->save_all(dense_->parameters.output_dir.c_str());
	std::cout << "Done!" << std::endl;
	ros::Duration(1.0).sleep();
}

**Comment by Warsin** [comment]
@taihup this is log what i meet

 [RUNNING]  Bag Time: 1385799367.648178   Duration: 28.087016 / 28.108940       
 [RUNNING]  Bag Time: 1385799367.658286   Duration: 28.097124 / 28.108940       
 [RUNNING]  Bag Time: 1385799367.668401   Duration: 28.107239 / 28.108940               
Done.
root@ubuntu-virtual-machine:/usr/src/dense_sptam# rosnode kill -a
killing:
 * /dense
 * /nodelet_manager
 * /rosout
 * /sptam
 * /sptam_path
 * /start_broadcaster
[ WARN] [1529926160.100141618, 1385799367.669410815]: Shutdown request received. Reason: [user request]
[ WARN] [1529926160.101862879, 1385799367.669410815]: Shutdown request received.
[ WARN] [1529926160.101994488, 1385799367.669410815]: Reason given for shutdown: [user request]
[ WARN] [1529926160.105003662, 1385799367.669410815]: Shutdown request received. Reason: [user request]
killed
[ INFO] [1529926160.118841885, 1385799367.669410815]: Unloading nodelet /sptam from manager nodelet_manager
[ INFO] [1529926160.123604531, 1385799367.669410815]: Unloading nodelet /dense from manager nodelet_manager
root@ubuntu-virtual-machine:/usr/src/dense_sptam# [rosout-1] process has finished cleanly
log file: /root/.ros/log/4b26667e-786a-11e8-903a-000c2986eb06/rosout-1*.log
[start_broadcaster-2] process has finished cleanly
log file: /root/.ros/log/4b26667e-786a-11e8-903a-000c2986eb06/start_broadcaster-2*.log
[sptam_path-3] process has finished cleanly
log file: /root/.ros/log/4b26667e-786a-11e8-903a-000c2986eb06/sptam_path-3*.log
[rosout-1] restarting process
Starting DENSE node cleanup...
process[rosout-1]: started with pid [74126]
================================================================================REQUIRED process [nodelet_manager-4] has died!
process has died [pid 67653, exit code -11, cmd /opt/ros/kinetic/lib/nodelet/nodelet manager __name:=nodelet_manager __log:=/root/.ros/log/4b26667e-786a-11e8-903a-000c2986eb06/nodelet_manager-4.log].
log file: /root/.ros/log/4b26667e-786a-11e8-903a-000c2986eb06/nodelet_manager-4*.log
Initiating shutdown!
================================================================================
[dense-6] killing on exit
[sptam-5] killing on exit
[nodelet_manager-4] killing on exit
[rosout-1] killing on exit
[master] killing on exit
shutting down processing monitor...
... shutting down processing monitor complete
done

**Comment by taihup** [comment]
mmm as I remember ctrl-c  was the way to kill the dense-sptam node... I
will try to check it when I have a moment. There could be bugs as always ;S

If you make dense-sptam works, remember to comment the solution so it can
help to others.

**Comment by wutongyu98** [comment]
> but when i run
> roslaunch src/dense-sptam/launch/kitti.launch &
> rosbag play --clock --rate 0.1 kitti_04.bag
> i meet this:
> auto-starting new master
> process[master]: started with pid [85]
> ROS_MASTER_URI=http://127.0.0.1:11311
> 
> setting /run_id to b1117b30-7532-11e8-a2c8-000c2986eb06
> process[rosout-1]: started with pid [98]
> started core service [/rosout]
> process[start_broadcaster-2]: started with pid [118]
> process[sptam_path-3]: started with pid [119]
> process[nodelet_manager-4]: started with pid [137]
> [ INFO] [1529572125.428744450]: Initializing nodelet with 4 worker threads.
> process[sptam-5]: started with pid [161]
> [ INFO] [1529572125.893189418]: Loading nodelet /sptam of type sptam/sptam_nodelet to manager nodelet_manager with the following remappings:
> [ INFO] [1529572125.893319381]: /stereo/left/camera_info -> /kitti_stereo/left/camera_info
> [ INFO] [1529572125.893357602]: /stereo/left/image_rect -> /kitti_stereo/left/image_rect
> [ INFO] [1529572125.893587453]: /stereo/right/camera_info -> /kitti_stereo/right/camera_info
> [ INFO] [1529572125.893632843]: /stereo/right/image_rect -> /kitti_stereo/right/image_rect
> crossCheck: 0
> OpenCV: threads set to -1
> nfeatures: 1000
> qualityLevel: 0.01
> minDistance: 15
> blockSize: 3
> useHarrisDetector: 0
> k: 0.04
> nfeatures: 1000
> qualityLevel: 0.01
> minDistance: 15
> blockSize: 3
> useHarrisDetector: 0
> k: 0.04
> bytes: 32
> use_orientation: 0
> bytes: 32
> use_orientation: 0
> process[dense-6]: started with pid [188]
> [ INFO] [1529572126.490738690]: Loading nodelet /dense of type dense/denseNodelet to manager nodelet_manager with the following remappings:
> [ INFO] [1529572126.490883237]: /dense/keyframes -> /sptam/keyframes
> [ INFO] [1529572126.782521508]: S-PTAM stereo node initialized.
> [ INFO] [1529572127.322514652]: DENSE node initialized.
> [ INFO] [1529572130.203967090]: Opening kitti_04.bag
> 
> Waiting 0.2 seconds after advertising topics... done.
> 
> Hit space to toggle paused, or 's' to step.
> [ INFO] [1529572130.735223612, 1385799339.661297826]: init calib8.108940
> [ INFO] [1529572130.735660508, 1385799339.661297826]: baseline: 0.537151
> [ INFO] [1529572130.785293238, 1385799339.661297826]: Trying to intialize map...
> [ INFO] [1529572130.842278877, 1385799339.661297826]: Map initialized with 137 points.
> [RUNNING] Bag Time: 1385799367.148107 Duration: 27.586945 / 28.108940
> Done.
> it seems that just play the bag and do nothing.

I have exactly same problem, the program only play the bag, but no results generated. Have you get this problem solved? Thanks,

**Comment by taihup** [comment]
Did you try to see the point cloud and pose in rviz?

**Comment by wutongyu98** [comment]
> Did you try to see the point cloud and pose in rviz?

I did, but the rviz didn't see any map. Even though I see many map topic being published, such as /dense/dense_cloud, /sptam/local_map. So I doubt there is no results were generated. To verify my guess, I add some debugging in the base_driver.cpp. The bold code were the test code. Unfortunately, it only print "the point is 0" during the whole programming running.
 // create map file
  {
    std::cout << "saving map cloud..." << std::endl;
    std::ofstream out("map cloud.dat");
    **int i = 0;
    std::cout << "the point is " << i++ << std::endl;**
    for ( const auto& point : sptam_->GetMap().getMapPoints() ) {
		**std::cout << "the point is " << i++ << std::endl;**
      out << point->GetPosition().x() << " " << point->GetPosition().y() <<  " " << point->GetPosition().z() << std::endl;
  }
    out.close();
  }

I think, the program might not get the image bag. Maybe my rosbag is not in the right format. Could you send me a copy of your ros bag for sequence 6 you have already succeed during your test? My email is jxh831027@gmail.com. Thanks,

**Comment by wutongyu98** [comment]
> Try download the datasets from [kitti_00.bag](http://fs01.cifasis-conicet.gov.ar:90/~pire/datasets/KITTI/bags/kitti_00.bag). It is possible to download the rest of the kitti sequences from [KITTI rosbags link](http://fs01.cifasis-conicet.gov.ar:90/~pire/datasets/KITTI/bags/). The KITTI dataset links are mentioned in the original S-PTAM repository. Check: https://github.com/lrse/sptam.

Thanks taihup. The sptam works with the new rosbag. But the dense point didn't work. I looked into the code and found the "if (!dense_)" was executed in "void dense::denseInterface::cb_keyframes_path(const nav_msgs::PathConstPtr& path)". Do you have an idea what's the problem? Thanks

**Comment by taihup** [comment]
What is the error? Just to be sure, Did you create the clouds repository in ~/.ros folder? Did you run the commands `roscore`,  `rosparam set use_sim_time true` and `rosbag play --clock <rosbag>`?

**Comment by wutongyu98** [comment]
> What is the error? Just to be sure, Did you create the clouds repository in ~/.ros folder? Did you run the commands `roscore`, `rosparam set use_sim_time true` and `rosbag play --clock <rosbag>`?
Yes, I did. But no difference, still produce no dense point. When I try to load the topic " sub_img_r_.subscribe(nhp, "/keyframe/right/image_rect", 1);" in **denseInterface.cpp**, I found there were no image display. So I change **keyframe** with **kitti_stereo**, the code can produce dense map, but it crash after a very short while (~ 2s). The good thing is that this change really produce dense map, which can be load from rviz. 

Summary, sptam can work well, dense map only worked for 2s after I change   **keyframe** with **kitti_stereo** in the denseInterface.cpp
        sub_img_l_.subscribe(nhp, "/**keyframe**/left/image_rect", 1);
	sub_info_l_.subscribe(nhp, "/**keyframe**/left/camera_info", 1);
	sub_img_r_.subscribe(nhp, "/**keyframe**/right/image_rect", 1);
	sub_info_r_.subscribe(nhp, "/**keyframe**/right/camera_info", 1);

**Comment by taihup** [comment]
Run the rosbag with a lower rate: 

rosbag play --clock --rate 0.2 <rosbag>

**Comment by wutongyu98** [comment]
> Run the rosbag with a lower rate:
> 
> rosbag play --clock --rate 0.2 rosbag_kitti

Just confirmed that this crash is caused by the parameter **local_area_size**. When I set this size 10, it will run 11 iteration, when I set it 100, it will run 101 iteration. Then crash. I also check the erase operation in **void PointCloudQueue::push_local_area(PointCloudEntry::Ptr entry)** in PointCloudQueue.cpp, it works well. Any suggestion, why this crash happen? thanks

**Comment by taihup** [comment]
I do not know about this crash. In my experience, unexpected crashes of S-PTAM and Dense-sptam are related with the vectorization part of Eigen library or with the use of libraries compiled with different versions of Eigen library.

Are you using the docker right or you are installing the system in your own ubuntu?

**Comment by taihup** [comment]
Yes there are several parameters that you can tune in the yaml configuration files, for instance you can increase the local_area_size =20 parameter to fuse information from the last 20 keyframes.

**Comment by LuoMing2** [comment]
How did you got the dense map? I also got 

this question when I change the "local area size"

> > Run the rosbag with a lower rate:
> > rosbag play --clock --rate 0.2 rosbag_kitti
> 
> Just confirmed that this crash is caused by the parameter **local_area_size**. When I set this size 10, it will run 11 iteration, when I set it 100, it will run 101 iteration. Then crash. I also check the erase operation in **void PointCloudQueue::push_local_area(PointCloudEntry::Ptr entry)** in PointCloudQueue.cpp, it works well. It looks the some index was not update after begin entry of point_cloud_ was erased. Any suggestion about this crash? thanks

**Comment by taihup** [comment]
@LuoMing2, have you checkted this answer: [https://github.com/CIFASIS/dense-sptam/issues/3#issuecomment-399058740](https://github.com/CIFASIS/dense-sptam/issues/3#issuecomment-399058740)

**Comment by LuoMing2** [comment]
> @LuoMing2, have you checkted this comment: [#3 (comment)](https://github.com/CIFASIS/dense-sptam/issues/3#issuecomment-399058740)?

yep, I followed wutongyu98's solution, but the program still broken when its came to iteration = 10, as the local_area_size = 10.

**Comment by YznMur** [comment]
@taihup @LuoMing2 Hi
I am facing the same problem here, did you manage to solve it?


> > @LuoMing2, have you checkted this comment: [#3 (comment)](https://github.com/CIFASIS/dense-sptam/issues/3#issuecomment-399058740)?
> 
> yep, I followed wutongyu98's solution: [#3 (comment)](https://github.com/CIFASIS/dense-sptam/issues/3#issuecomment-438044074). But the program still broken when its came to iteration = 10, as the local_area_size = 10. @wutongyu98
